File: src/frontend/src/api/notifications.ts
import { apiClient } from './client'
import type { NotificationRead, UnreadCountRead } from '../types/notification'

export const notificationsApi = {
  list: (limit = 50) =>
    apiClient.get<NotificationRead[]>('/notifications', { params: { limit } }).then((r) => r.data),
  unreadCount: () =>
    apiClient.get<UnreadCountRead>('/notifications/unread-count').then((r) => r.data),
  markRead: (id: string) =>
    apiClient.patch(`/notifications/${id}/read`),
  markAllRead: () =>
    apiClient.post('/notifications/read-all'),
}

File: src/backend/app/api/notifications.py
import uuid

from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.dependencies import get_current_user
from app.models.user import User
from app.schemas.notification import NotificationRead, UnreadCountRead
from app.services.notification import NotificationService

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("", response_model=list[NotificationRead])
async def list_notifications(
    limit: int = 50,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return await NotificationService(db).list_for_user(current_user.id, limit)


@router.get("/unread-count", response_model=UnreadCountRead)
async def unread_count(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    count = await NotificationService(db).unread_count(current_user.id)
    return {"count": count}


@router.patch("/{notification_id}/read", status_code=status.HTTP_204_NO_CONTENT)
async def mark_read(
    notification_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await NotificationService(db).mark_read(notification_id, current_user.id)


@router.post("/read-all", status_code=status.HTTP_204_NO_CONTENT)
async def mark_all_read(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await NotificationService(db).mark_all_read(current_user.id)
